app/models/binary_blob/purging.rb

```diff
@@ -13,7 +13,7 @@ def purge_window_size
       end
 
       def purge_scope(_older_than = nil)
-        where(:resource => nil)
+        where(:resource_type => 'MiqQueue').where.not('resource_id in (select id from miq_queue)')
```

Comment:
hm, this would explode, given ```select id from miq_queue``` would unpack the whole miq queue (that can go to milions)

```
refresh_queue_items = MiqQueue.where(:method_name => 'refresh') # rails is smart enough to add .select(:id)
where(:resource_type => 'MiqQueue').where.not(:resource_id => refresh_queue_items)
```

I think ```:method_name => 'refresh'``` might be enough to fetch existing refresh MiqQueue items, or maybe more from https://github.com/Ladas/manageiq/blob/acf2ad8e29b9e70858fc99e22614d2112312066c/app/models/ems_refresh.rb#L162 ?

Is purging Zone agnostic? Or we should limit it by zone, if each zone has its own purger?

Reply:
query looks fast enough in hundred k world
```
2.4.2 :019 > BinaryBlob.where(:resource_type => 'MiqQueue').where.not(:resource_id => refresh_queue_items).count
   (92.6ms)  SELECT COUNT(*) FROM "binary_blobs" WHERE "binary_blobs"."resource_type" = $1 AND ("binary_blobs"."resource_id" NOT IN (SELECT "miq_queue"."id" FROM "miq_queue" WHERE "miq_queue"."method_name" = $2))  [["resource_type", "MiqQueue"], ["method_name", "refresh"]]
  Query Trace > lib/extensions/ar_virtual.rb:741:in `calculate'
 => 351020
```

Reply:
Do we only care about purging blobs related to `MiqQueue`? If not, you can use the polymorphic orphan purging I added in https://github.com/ManageIQ/manageiq/pull/16754

It [uses a join](https://github.com/ManageIQ/manageiq/pull/16754#issuecomment-355649210) so it should be a bit more efficient than the subquery you have here. Also it won't pull back the ids to the server at all.

Reply:
@carbonin 👍 looks much better, I'll switch to that when yours is merged